python/contracts.py

```diff
@@ -0,0 +1,387 @@
+        if p.kind == p.POSITIONAL_OR_KEYWORD and p.default == p.empty
+    }
+    return params_args, mandatory, unexpected
+
+
+def check_requirements(params_args, mandatory, unexpected, contracts):
+    """
+    Check whether `params_args`, `mandatory` and `unexpected` meet the requirements of contracts.
+
+    :param params_args: dict of parameters to arguments
+    :param mandatory: set of required positional parameters
+    :param unexpected: unexpected keyword parameters
+    :param contracts: dict of parameters to requirements (a `Requirement` object)
+    :return: parameters to arguments, required positional parameters and unexpected keyword parameters
+             tuple(dict{param: arg, ...}, set{param, ...}, set{param, ...})
+    :raise: raise Diagnostics if the check failed
+    """
+    diagnostics = Diagnostics()
+    for param, arg in filter(lambda t: t[0] in contracts, params_args.items()):
+        # Checking violations
+        if not contracts[param](arg):
+            diagnostics[param] = str(contracts[param]), arg, param in mandatory
+    for param in mandatory - params_args.keys():
+        # Checking missing arguments
+        diagnostics[param] = str(
+            contracts[param]) if param in contracts else ''
+    for param in unexpected:
+        # Checking unexpected arguments
+        diagnostics[param] = "UNEXPECTED"
+    if diagnostics:
+        raise diagnostics
+
+
+def check_requirements_for_existed(func, kwargs, **contracts):
+    """
+    Check whether `kwargs` meet the requirements of `func`'s `contracts`.
+
+    :param func: a function or callable object to be extracted
+    :param kwargs: dict of the form {param: arg, ...}, variadic keyword arguments
+    :param contracts: variadic keyword arguments of parameter names to `Requirement`s
+    :return: parameters to arguments, required positional parameters and unexpected keyword parameters
+             tuple(dict{param: arg, ...}, set{param, ...}, set{param, ...})
+    :raise: raise Diagnostics if the check failed
+    """
+    check_requirements(*extract_params_args(func, [], kwargs), contracts)
+
+
+# This is an example of using `contracts` to design a framework to
+# check various callables
+class Contracts(object):
+    def __init__(self):
+        self._contracts = {}
+
+    def add_requirements(self, func, *aliases, **contracts):
+        for alias in list(aliases) + [func]:
+            self._contracts[alias] = contracts
+
+    def check_requirements(self, func, kwargs):
+        if func in self._contracts:
+            check_requirements_for_existed(func, kwargs,
+                                           **self._contracts[func])
+
+
+if __name__ == '__main__':
+    import tensorflow as tf
+    import xgboost as xgb
+    my_contracts = Contracts()
+    my_contracts.add_requirements(tf.estimator.DNNClassifier,
+                                  hidden_units=Type([PositiveInt]),
+                                  n_classes=Int & GreaterEqual(2),
```

Comment:
Do we actually need `AND`?
From the prototype, we can express `0<x<10` by three ways:
``` python
Int & Greater(0) & Less(10)
Int & PositiveInt & Less(10)
Int & Between(0, 10)
```
Is that what we want? I think that may output different diagnostic messages. And maybe we can use API instead of overload operators can be more clear?

``` python
x=Int(left=0, right=10, is_left_closed=False, is_right_closed=False)
```

So, I think we don't need `AND`, just `OR` if enough.

Reply:
We need `AND`, or we have to define something like:
```python
x = Int(left=0, right=10, is_left_closed=False, is_right_closed=False)
y = Float(left=0.0, right=10.0, is_left_closed=False, is_right_closed=False)
```
Tha't not pythonic because python support `generic programming` , we don't need to implement two functions/classes with the same implementation.
So we have to define
```python
x = Number([int], left=0, right=10, is_left_closed=False, is_right_closed=False)
y = Number([int, float], left=0, right=10, is_left_closed=False, is_right_closed=False)
```
This is tediou and error-prone, and the usecase of `y` will be very common in `XGBoost`. Moreover, the implementation of `Number` has to distinguish many cases to generate a friendly error message.This is also anti-pythonic and error-prone:
```python
def Number(v, typ, left, right, is_left_closed, is_right_closed):
    assert not (left = None and right == None),   "the range is empty"
    assert typ in (int, float)
    assert something else...

    if left = None:
        if is_right_closed and v > right
            raise Diagnostics(f"should be {typ.__name__} that <= {}")
        elif is_right_closed and v >= right: 
            raise Diagnostics(f"should be {typ.__name__} that < {}")
    if right = None:
        # do something similar
    if is_right_closed and not is_left_closed:
    # ...
```
Now that we're writing python code, we should write them in a `pythonic` way, not in the `golang`'s way. The [Google Python Style](http://google.github.io/styleguide/pyguide.html) doesn't forbid operator overloading (Actually it doesn't say a word about operator overloading).  If we refer to the [Google C++ Style](https://google.github.io/styleguide/cppguide.html) described: 
- Define overloaded operators only if their meaning is obvious, unsurprising, and consistent with the corresponding built-in operators. For example, **use | as a bitwise- or logical-or**, not as a shell-style pipe.
- **Don't go out of your way to avoid defining operator overloads.** For example, prefer to define `==`, `=`, and `<<`, rather than `Equals()`, `CopyFrom()`, and `PrintTo()`.